appinventor/components/src/com/google/appinventor/components/runtime/ChartDataBase.java

```diff
@@ -0,0 +1,957 @@
+import com.google.appinventor.components.annotations.SimpleFunction;
+import com.google.appinventor.components.annotations.SimpleObject;
+import com.google.appinventor.components.annotations.SimpleProperty;
+import com.google.appinventor.components.common.ComponentConstants;
+import com.google.appinventor.components.common.PropertyTypeConstants;
+import com.google.appinventor.components.runtime.util.CsvUtil;
+import com.google.appinventor.components.runtime.util.OnBeforeInitializeListener;
+import com.google.appinventor.components.runtime.util.YailList;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.*;
+
+@SimpleObject
+public abstract class ChartDataBase implements Component, OnBeforeInitializeListener, ChartDataSourceChangeListener,
+    ChartDataSourceGetValueListener {
+    protected Chart container;
+    protected ChartDataModel chartDataModel;
+
+    /* Used to queue & execute asynchronous tasks while ensuring
+     * order of method execution (ExecutorService should be a Single Thread runner)
+     * In the case of methods which return values and where
+     * the result depends on the state of the data, blocking get
+     * calls are used to ensure that all the previous async tasks
+     * finish before the data is returned. */
+    protected ExecutorService threadRunner;
+
+    // Properties used in Designer to import from DataFile.
+    // Represents the names of the columns to use,
+    // where each index corresponds to a single dimension.
+    protected List<String> dataFileColumns;
+
+    // Properties used in Designer to import from Web components.
+    // Represents the names of the columns to use,
+    // where each index corresponds to a single dimension.
+    protected List<String> webColumns;
+
+    // Property used in Designer to import from a Data Source.
+    // Represents the key value of the value to use from the
+    // attached Data Source.
+    protected String dataSourceValue;
+
+    private String label;
+    private int color;
+    private YailList colors;
+    private int pointShape;
+    private int lineType;
+
+    private ChartDataSource dataSource; // Attached Chart Data Source
+
+    // Currently imported observed Data Source value. This has to be
+    // kept track of in order to remove old entries whenever the
+    // value is updated.
+    private Object currentDataSourceValue;
+
+    private String elements; // Elements Designer property
+
+    private boolean initialized = false; // Keep track whether the Screen has already been initialized
+
+    private int t = 0;
+
+    /**
+     * Creates a new Chart Data component.
+     */
+    protected ChartDataBase(Chart chartContainer) {
+        this.container = chartContainer;
+        chartContainer.addDataComponent(this);
+
+        // Set default properties and instantiate Chart Data Model
+        initChartData();
+        DataSourceValue("");
```

Comment:
@ewpatton InitChartData has a difference purpose in the sense that it only initializes chart specifics and the Chart Data Model itself. I'm a bit worried about putting the Source property in there if ever an issue pops up with the ordering events in future changes. Namely, it could be the case that the Source property would be unset when the type of the Chart changes. For a bit more context, see the `reinitializeChart` method in the Chart (component) class.